### ISSUE TITLE
<string>: Potentially missing terminator when ASAN is enabled.

### ISSUE THREAD

**Issue by W4RH4WK** [microsoft/STL#3188]
I am not 100% certain about this one, but while debugging one of our tools (with address sanitizer enabled) I discovered an `std::string` with a non-null-terminated payload. This is problematic as `c_str()` will just return a pointer to the internal buffer:

https://github.com/microsoft/STL/blob/c873cf01323f2f37ce3021f344d905dc84411fe3/stl/inc/xstring#L4254-L4256

https://github.com/microsoft/STL/blob/c873cf01323f2f37ce3021f344d905dc84411fe3/stl/inc/xstring#L2290-L2297

Tracking down the construction, the problematic code is here:

https://github.com/microsoft/STL/blob/c873cf01323f2f37ce3021f344d905dc84411fe3/stl/inc/xstring#L2754-L2758

Instead of moving the whole buffer (including the terminator) from the source, this moves only `_Count` characters, which does not include the terminator. Subsequent code does/did not add the required terminator.

**Comment by miscco** [comment]
@BillyONeal see I fixed your off by one to introduce another one.

**Comment by CaseyCarter** [comment]
We believe this bug in version 17.3 was fixed in 17.4p1 by #1735. @W4RH4WK does this still reproduce for you in 17.4?

**Comment by CaseyCarter** [comment]
> I encountered the issue on 17.3.4. I'll report back once I get around to checking out 17.4.

I should point out that ASan has some other issues with `std::string` that caused us to disable annotations; they'll be back after #1764 lands, probably in 17.6. You can define `_DISABLE_STRING_ANNOTATION` if you want to use ASan and avoid ASan integration with `std::string` in 17.3.

**Comment by StephanTLavavej** [comment]
We're going to go ahead and close this issue as the fix has been merged into the microsoft/STL GitHub repo and the MSVC-internal repo. Thanks again for the report! Please open a new issue if you encounter problems with 17.4 (which will be released for production in the near future although we can't say exactly when).
